src/python/dxpy/sugar/transfers.py

```diff
@@ -0,0 +1,952 @@
+import logging
+import multiprocessing
+import os
+import re
+import sys
+
+import dxpy
+from dxpy.compat import basestring, makedirs
+from dxpy.utils.resolver import data_obj_pattern
+from dxpy.sugar import processing as proc
+
+
+# Helpers - these belong in other modules
+
+import tempfile
+
+
+@contextlib.contextmanager
+def tmpfile(*args, **kwargs):
+    """
+    Create a temporary file, yield it, and delete it before returning.
+
+    Yields:
+        A path to a temporary file.
+
+    Notes:
+        This method is needed distinct from :class:`tempfile.TemporaryFile` in the
+        case where python needs to write to the file and then a subprocess needs
+        to read from the file.
+    """
+    path = tempfile.mkstemp(*args, **kwargs)[1]
+    try:
+        yield path
+    finally:
+        if os.path.exists(path):
+            os.remove(path)
+
+# End helpers
+
+
+LOG = logging.getLogger()
+SPECIAL_RE = re.compile(r"[^\w.]")
+MAX_READ_SIZE = (1024 * 1024 * 1024 * 2) - 1
+"""Maximum number of bytes that can be read from a file at once. The limit here is
+due to a known bug in some versions of python on macOS:
+https://bugs.python.org/issue24658.
+"""
+
+
+def upload_file(filename, skip_compress=False, **kwargs):
+    """
+    Upload a file, optionally compressing it.
+
+    Args:
+        filename (str): The file to upload.
+        skip_compress (bool): Whether to skip gzip compression.
+        kwargs: Additional kwargs to the upload function.
+
+    Returns:
+        A dxlink or DXFile.
+    """
+    if skip_compress:
+        return simple_upload_file(filename, **kwargs)
+    else:
+        return compress_and_upload_file(filename, **kwargs)
+
+
+def simple_upload_file(
+    filename,
+    remote_filename=None,
+    remote_folder="/",
+    return_handler=False,
+    project=None,
+    wait_on_close=False,
+):
+    """
+    Upload a file and return a link.
+
+    Args:
+        filename (str): Local filename.
+        remote_filename (str): Optional, remote filename to upload file to.
+        remote_folder (str): Optional, remote folder to upload file to.
+        return_handler (bool): Whether to return a DXFile handler.
+        project (str): The project ID to upload to, if not the currently selected
+            project.
+        wait_on_close (bool): Whether to block until the file has closed.
+
+    Returns:
+        DNAnexus link or DXFile pointing to the uploaded file (depending on
+        the value of `return_handler`).
+    """
+    kwargs = {"wait_on_close": wait_on_close}
+
+    if remote_filename:
+        kwargs["name"] = remote_filename
+    if not remote_folder:
+        remote_folder = "/"
+    elif not remote_folder.startswith("/"):
+        remote_folder = "/{}".format(remote_folder)
+    kwargs["folder"] = remote_folder
+    if project:
+        kwargs["project"] = project
+    LOG.info("Uploading file %s to %s", filename, remote_filename or filename)
+    handler = dxpy.upload_local_file(filename, **kwargs)
```

Comment:
As much as I like `remote_filename` and `remote_folder` as being more descriptive, I think the passed-in kwargs should be consistent with other dxpy functions, so should just be `name` and `folder` for the remote file/folder locations. To avoid confusion you could replace `filename` with `local_path`

Reply:
Done